fdbserver/ptxn/TLogPeekCursor.actor.h

```diff
@@ -0,0 +1,316 @@
+	iterator begin() { return iterator(this, false); }
+
+	// Returns an iterator that represents the end of the data
+	const iterator& end() const { return endIterator; }
+
+protected:
+	// Checks if there is any mutations remotely
+	virtual Future<bool> remoteMoreAvailableImpl() = 0;
+
+	// Steps the local cursor
+	virtual void nextImpl() = 0;
+
+	// Gets the message
+	virtual const VersionSubsequenceMutation& getImpl() const = 0;
+
+	// Checks if any remaining mutations
+	virtual bool hasRemainingImpl() const = 0;
+
+private:
+	// The iterator that represents the end of the unserialized data
+	const iterator endIterator;
+};
+
+// Connect to given TLog server(s) and peeks for mutations with a given TeamID
+class StorageTeamPeekCursor : public PeekCursorBase {
+	const StorageTeamID storageTeamID;
+	std::vector<TLogInterfaceBase*> pTLogInterfaces;
+
+	// The arena used to store incoming serialized data, if not nullptr, TLogPeekReply arenas will be attached to this
+	// arena, enables the access of deserialized data even the cursor is destroyed.
+	Arena* pAttachArena;
+	TLogStorageServerMessageDeserializer deserializer;
+	TLogStorageServerMessageDeserializer::iterator deserializerIter;
+
+	// Returns the begin verion for the cursor. The cursor will start from the begin version.
+	const Version& getBeginVersion() const;
+
+	// The version the cursor starts
+	const Version beginVersion;
+
+	// Last version processed
+	Version lastVersion;
+
+public:
+	// version_ is the version the cursor starts with
+	// storageTeamID_ is the storageTeamID
+	// pTLogInterface_ is the interface to the specific TLog server
+	// pArena_ is used to store the serialized data for further use, e.g. making MutationRefs still available after the
+	// cursor is destroyed. If pArena_ is nullptr, any reference to the peeked data will be invalidated after the cursor
+	// is destructed.
+	StorageTeamPeekCursor(const Version& version_,
+	                     const StorageTeamID& storageTeamID_,
+	                     TLogInterfaceBase* pTLogInterface_,
+	                     Arena* pArena_ = nullptr);
+
+	StorageTeamPeekCursor(const Version& version_,
+	                     const StorageTeamID& storageTeamID_,
+	                     const std::vector<TLogInterfaceBase*>& pTLogInterfaces_,
+	                     Arena* arena_ = nullptr);
+
+	const StorageTeamID& getTeamID() const;
+
+	// Returns the last version being pulled
+	const Version& getLastVersion() const;
+
+protected:
+	virtual Future<bool> remoteMoreAvailableImpl() override;
+	virtual void nextImpl() override;
+	virtual const VersionSubsequenceMutation& getImpl() const override;
+	virtual bool hasRemainingImpl() const override;
+};
+
+using CursorContainer = std::list<std::unique_ptr<PeekCursorBase>>;
+
+// This class defines the index of a cursor. For a given mutation, it has an unique cursor index. The index is totally
+// ordered.
+struct IndexedCursor {
```

Comment:
Can this be defined inside `PeekCursorBase`? It's a little strange to define this as a separate data type.

Reply:
Sure. Actually putting it into `PeekCursorBase` does not make too much sense as `PeekCursorBase` should not know anything about a cursor container, so I have put it into `MergedPeekCursorBase`.